src/setup.rs

```diff
@@ -140,5 +150,198 @@ where
     signal_handler::wait_for_signal();
 
     // Stop server
-    server.stop();
+    server.stop().await;
+}
+
+async fn build_in_cluster_mode<Q: Executor + 'static>(
+    config: &Config,
+    builder: Builder<Q>,
+    runtimes: &EngineRuntimes,
+    engine_proxy: TableEngineRef,
+) -> Builder<Q> {
+    let meta_client = meta_impl::build_meta_client(
+        config.cluster.meta_client.clone(),
+        config.cluster.node.clone(),
+        runtimes.meta_runtime.clone(),
+    )
+    .unwrap();
+
+    let catalog_manager = {
+        let schema_id_alloc = SchemaIdAllocWrapper(meta_client.clone());
+        let table_id_alloc = TableIdAllocWrapper(meta_client.clone());
+        Arc::new(volatile::ManagerImpl::new(schema_id_alloc, table_id_alloc).await)
+    };
+
+    let cluster = {
+        let table_manipulator = Arc::new(TableManipulatorImpl {
+            catalog_manager: catalog_manager.clone(),
+            table_engine: engine_proxy,
+        });
+        let cluster_impl = ClusterImpl::new(
+            meta_client,
+            table_manipulator,
+            config.cluster.clone(),
+            runtimes.meta_runtime.clone(),
+        )
+        .unwrap();
+        Arc::new(cluster_impl)
+    };
+
+    builder.catalog_manager(catalog_manager).cluster(cluster)
+}
+
+async fn build_in_standalone_mode<Q: Executor + 'static>(
+    builder: Builder<Q>,
+    table_engine: TableEngineRef,
+    engine_proxy: TableEngineRef,
+) -> Builder<Q> {
+    let table_based_manager = TableBasedManager::new(table_engine, engine_proxy.clone())
+        .await
+        .unwrap_or_else(|e| {
+            panic!("Failed to create catalog manager, err:{}", e);
+        });
+
+    // Create catalog manager, use analytic table as backend
+    let catalog_manager = Arc::new(CatalogManagerImpl::new(Arc::new(table_based_manager)));
+    builder.catalog_manager(catalog_manager)
+}
+
+struct SchemaIdAllocWrapper(MetaClientRef);
```

Comment:
How about moving `SchemaIdAllocWrapper`, `TableIdAllocWrapper` and `TableManipulatorImpl` other places?